### PR TITLE
fix: 301 and 302 change method to GET

### PR DESCRIPTION
**Pull request by DTrombett**
## This relates to...

Didn't open an issue before this PR so sorry if this is expected but I could not find any discussion about it.
The redirect handler doesn't change the method to `GET` when the status code is 301/302 and the original method is `POST` despite being the standard in fetch spec and browser behavior. Also the HTTP standard mentions this too:
> Note: For historical reasons, a user agent MAY change the request method from POST to GET for the subsequent request. If this behavior is undesired, the 307 (Temporary Redirect) status code can be used instead.

## Changes

Method is changed to `GET` and body is removed when a 301/302 response code is received, aligning with fetch and browser behavior
The tests were updated to reflect this change and the ones specifically about request body/method were updated to use `PUT` instead

## Status

- [x] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [x] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [x] Review ready
- [x] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Review by ronag — APPROVED** [review]

File: lib/handler/redirect-handler.js [hunk 1]
```diff
@@ -132,6 +132,14 @@ class RedirectHandler {
     this.opts.maxRedirections = 0
     this.opts.query = null
 
+    // https://tools.ietf.org/html/rfc7231#section-6.4.2
+    // https://fetch.spec.whatwg.org/#http-redirect-fetch
+    // In case of HTTP 301 or 302 with POST, change the method to GET
+    if ((statusCode === 301 || statusCode === 302) && this.opts.method === 'POST') {
+      this.opts.method = 'GET'
+      this.opts.body = null
```
Comment (ronag) [inline lib/handler/redirect-handler.js:140]: If body is a stream it needs to be destroyed.

**Review by ronag — CHANGES_REQUESTED** [review]

File: lib/handler/redirect-handler.js (hunk 1 above)
Comment (DTrombett) [reply lib/handler/redirect-handler.js:140]: Do we need to do update the HTTP 303 case as well?

File: lib/handler/redirect-handler.js [hunk 2]
```diff
@@ -132,10 +132,20 @@ class RedirectHandler {
     this.opts.maxRedirections = 0
     this.opts.query = null
 
+    // https://tools.ietf.org/html/rfc7231#section-6.4.2
+    // https://fetch.spec.whatwg.org/#http-redirect-fetch
+    // In case of HTTP 301 or 302 with POST, change the method to GET
+    if ((statusCode === 301 || statusCode === 302) && this.opts.method === 'POST') {
+      this.opts.method = 'GET'
+      util.destroy(this.opts.body)
```
Comment (ronag) [inline lib/handler/redirect-handler.js:140]:
```suggestion
      util.destroy(this.opts.body.on('error', () => {})
```

**Review by ronag — APPROVED** [review]

File: lib/handler/redirect-handler.js (hunk 2 above)
Comment (DTrombett) [reply lib/handler/redirect-handler.js]: Maybe is better to use `util.isStream()`?

**Comment by Uzlopak** [comment]
semver major?

**Pull request merged**